### DISCUSSION TITLE
cannot clearErrors manually before onSubmit

### DISCUSSION THREAD

**Discussion by aryaniyaps** [react-hook-form/react-hook-form#4924] [Q&A] [answered]
while setting errors using the `setError` method, a `ClearErrors` method needs to be called somewhere in order to submit the form.  (#4913)
![](https://user-images.githubusercontent.com/10513364/115816847-017f1c80-a43d-11eb-99ce-0adfd2859197.png)
however, I have a doubt and cannot figure out where to call this method.
But it has to be before the `onSubmit` method is called.

can someone please help me with this?

**Comment by JamesWong1999** [comment]
could you invoke `clearErrors` with input `onChange`?

**Reply by aryaniyaps** [reply]
Thanks for your response. I am yet to get my hands on the pc to try it,but it seems like it's good idea

**Comment by aryaniyaps** [comment]
```js
// FormField
export const FormField = props => {
  // access context from FormProvider
  const { register, clearErrors, formState: { errors } } = useFormContext();
  const field_errors = get(errors, props.name);
  
  const handleChange = e => {
    console.log('clearing')
    clearErrors(e.target.value);
    console.log('cleared')
  }

  return (
    <div className="formField-wrapper">
      <div className="header">
        <label className={`label ${field_errors && 'errors'}`} htmlFor={props.name}>{props.label}</label>
        {field_errors && field_errors.type === 'server' &&
          <label className="error-msg errors">{field_errors.message}</label>}
      </div>
      <TextField
        register={register}
        variant={field_errors ? `${props.variant} errors` : props.variant}
        type={props.type}
        name={props.name}
        defaultValue={props.defaultValue}
        value={props.value}
        placeholder={props.placeholder}
        onChange={handleChange}
        maxLength={props.maxLength}
        spellCheck={false}
      />
    </div>
  )
}
```
the event handler is not firing though
@JamesWong1999

**Reply by bluebill1049** [reply]
so why not set an error to an existing registered input? wouldn't that be easier to deal with?

```tsx
const { onChnage, ...rest } = register('test')

<input 
  onChange={(e) => {
    if (errors.customError) clearErrors(e.target.name);
    onChange(e); 
  }}
  ...rest
/>
```

**Reply by aryaniyaps** [reply]
@bluebill1049 
yes that works, but as I have figured out, I need to clear all the errors for the whole form somehow before submitting.
This is because, right now, only the fields that have changed are cleared.

If some other errors exists, still the form wont submit.

**Comment by apedretti** [comment, marked as answer]
Commenting here out of curiosity. Aryan, from what I understand your form has some validation and then _something_ happens and that validation is not needed anymore.

So you've got two alternatives:
- Have a dynamic validation.
- Clear the errors after that _something_ happened and before submit.

I'm not a big fan of clearing the errors if you can avoid it, so I've made you an example of the second alternative: [Conditional validation + useFormContext](https://codesandbox.io/s/react-hook-form-v7-form-context-forked-ckqe4). Try using that form with the console open, so you'll see the order of events.

Would that alternative work for you?

**Reply by aryaniyaps** [reply]
thanks a lot!
I guess that I can carry on with this.
appreciate your time.

**Reply by bluebill1049** [reply]
Thank your for the solution.
